scripts/py_matter_yamltests/matter_yamltests/yaml_loader.py

```diff
@@ -0,0 +1,235 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#
+#    Licensed under the Apache License, Version 2.0 (the 'License');
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an 'AS IS' BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from typing import Union
+
+from .errors import (TestStepError, TestStepGroupResponseError, TestStepInvalidTypeError, TestStepKeyError,
+                     TestStepNodeIdAndGroupIdError, TestStepValueAndValuesError, TestStepVerificationStandaloneError,
+                     TestStepWaitResponseError)
+from .fixes import add_yaml_support_for_scientific_notation_without_dot
+
+try:
+    from yaml import CSafeLoader as SafeLoader
+except:
+    from yaml import SafeLoader
+
+import yaml
+
+
+class YamlLoader:
+    """This class loads a file from the disk and validates that the content is a well formed yaml test."""
+
+    def load(self, yaml_file: str) -> tuple[str, Union[list, str], dict, list]:
+        name = ''
+        pics = None
+        config = {}
+        tests = []
+
+        if yaml_file:
+            with open(yaml_file) as f:
+                loader = SafeLoader
+                add_yaml_support_for_scientific_notation_without_dot(loader)
+                content = yaml.load(f, Loader=loader)
+
+                self.__check_content(content)
+
+                name = content.get('name', '')
+                pics = content.get('PICS')
+                config = content.get('config', {})
+                tests = content.get('tests', [])
+
+        return (name, pics, config, tests)
+
+    def __check_content(self, content):
+        schema = {
+            'name': str,
+            'PICS': (str, list),
+            'config': dict,
+            'tests': list,
+        }
+
+        try:
+            self.__check(content, schema)
+        except TestStepError as e:
+            if 'tests' in content:
+                # This is a top level error. The content of the tests section
+                # does not really matter here and dumping it may be counter-productive
+                # since it can be very long...
+                content['tests'] = 'Skipped...'
+            e.update_context(content, 0)
+            raise
+
+        tests = content.get('tests', [])
+        for step_index, step in enumerate(tests):
+            try:
+                self.__check_test_step(step)
```

Comment:
since the step is assumed to be ordered (to have popitem and reverse) should we somehow cast or create it as a https://docs.python.org/3/library/collections.html#collections.OrderedDict

?

Reply:
I guess python 3.7 or above also works based on docs. In that case some runtime assertion at least would be useful.

Reply:
maybe just:

```python
if sys.version_info <= (3,6):
  raise Exception("Need python >= 3.7 for dictionary order guarantees")
```

Reply:
I have added an assertion in `__init.py__`